src/core/namespace.c

```diff
@@ -3310,6 +3312,373 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+static int unpeel_get_fd(const char *mount_path, int *ret_fd) {
+        _cleanup_close_pair_ int pipe_fds[2] = EBADF_PAIR;
+        pid_t pid;
+        int r;
+
+        assert(mount_path);
+        assert(ret_fd);
+
+        r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, pipe_fds);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create socket pair: %m");
+
+        r = safe_fork("(sd-ns-unpeel)", FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                _cleanup_close_ int dir_fd = -EBADF;
+
+                pipe_fds[0] = safe_close(pipe_fds[0]);
+
+                /* Clone mount namespace here to unpeel without affecting live process */
+                r = detach_mount_namespace();
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to detach mount namespace: %m");
+                        _exit(EXIT_FAILURE);
+                }
+
+                /* Opportunistically unmount any overlay at this path */
+                r = umount_recursive(mount_path, MNT_DETACH);
```

Comment:
hmm, this is only ok if the relevant hierarchies (i.e. /etc, /usr/, /opt/) are not mount points anyway, which I think might not hold. at least two of these are often mount points.

hence i think this must be done in a more careful way, for example it might suffice to check if the mount refers to an overlayfs fs before unmounting it. If so, unmount it, if not assume we already reached the underlying fs.

Reply:
also, for the host side of things (i.e. the systemd-sysext tool) we recently added this: ddbddebe0cc5edcf3d0fbc82ba5850e45925e8ae

I sense we need to do something similar here: i.e. any mounts further down the tree probably not to be migrated to the new hierarchy too. people do use BindPaths= after all for some subdirs...